### PR TITLE
Flow: add loki.source.syslog component

### LINKED ISSUE #2505: `loki.source.syslog`: Listens over TCP or UDP for syslogs.

### PR DESCRIPTION
**Pull request by tpaschalis**
Signed-off-by: Paschalis Tsilias <paschalis.tsilias@grafana.com>

#### PR Description
The PR adds a loki.source.syslog component that spins up listeners for RFC5424-formatted syslog messages over TCP and UDP connections and forwards them to other `loki.*` components.

Each `listener` block in the component's Arguments will start a receiver. Promtail used the term `receiver`, but I chose to use a different one to avoid confusion with the receivers that tie together components. As with other cases,  entries are fanned out to each receiver passed in `ForwardTo` in succession.

Most of the PR size is due to bringing over Promtail's `syslog` package, so that we can use our own `loki.Entry` instance.

#### Which issue(s) this PR fixes
Fixes #2505.

#### Notes to the Reviewer

To test this component, I've used a configuration and a command like the ones presented below. I'm still not able to send messages over the UDP connection as I'm getting `msg="error parsing syslog stream" err="invalid or unsupported framing. first byte: 'X'"` even for valid RFC5424 entries.

```
# command for sending syslog messages
$ echo '<165>1 2022-04-08T22:14:10.001Z host1 app - id1 [custom@32473 exkey="1"] An application event log entry...' | nc 127.0.0.1 51893

# syslog.river
loki.source.syslog "local" {
	listener {
		address  = "127.0.0.1:51893"
		protocol = "tcp"

		labels = { component = "loki.source.syslog", protocol = "tcp" }
	}
	listener {
		address  = "127.0.0.1:51898"
		protocol = "udp"

		labels = { component = "loki.source.syslog", protocol = "udp"}
	}
	forward_to = [loki.write.grafanacloud.receiver]
}


loki.write "grafanacloud" {
  endpoint {
    url = "https://logs-prod-eu-west-0.grafana.net/loki/api/v1/push"
    http_client_config {
      basic_auth {
        username = "..."
        password = "eyJr="
      }
    }
  }
}
```

#### PR Checklist

- [x] CHANGELOG updated
- [X] Documentation added
- [X] Tests updated

### CODE REVIEW

File: docs/sources/flow/reference/components/loki.source.syslog.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+refers to a `tls_config` block defined inside a `config` block.
+
+[listener]: #listener-block
+[tls_config]: #tls_config-block
+
+### listener block
+
+The `listener` block defines the listen address and protocol where the listener
+will expect syslog messages to, as well as its behavior when receiving
+messages.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.syslog.md:60]:
```suggestion
The `listener` block defines the listen address and protocol where the listener
expects syslog messages to be sent to, as well as its behavior when receiving
messages.
```

Comment (karengermond) [reply docs/sources/flow/reference/components/loki.source.syslog.md:60]: Something seemed to be missing from this sentence, so I took a guess...

File: docs/sources/flow/reference/components/loki.source.syslog.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+The following arguments can be used to configure a `listener`. Only the
+`listen_address` field is required and any omitted fields take their default
+values.
+
+Name                     | Type          | Description | Default | Required
+------------------------ | ------------- | ----------- | ------- | --------
+`address`                | `string`      | The `<host:port>` address to listen on for syslog messages. | | yes
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.syslog.md:68]:
```suggestion
`address`                | `string`      | The `<host:port>` address to listen to for syslog messages. | | yes
```

File: docs/sources/flow/reference/components/loki.source.syslog.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+`protocol`               | `string`      | The protocol to listen on for syslog messages. Must be either `tcp` or `udp`. | `tcp` | no
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.syslog.md:69]:
```suggestion
`protocol`               | `string`      | The protocol to listen to for syslog messages. Must be either `tcp` or `udp`. | `tcp` | no
```

File: docs/sources/flow/reference/components/loki.source.syslog.md (hunk 2 above)
Comment (mattdurham) [inline docs/sources/flow/reference/components/loki.source.syslog.md:68]: Should this be `listen_address`?

File: docs/sources/flow/reference/components/loki.source.syslog.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+`idle_timeout`           | `duration`    | The idle timeout for tcp connections. | `"120s"` | no
+`label_structured_data`  | `bool`        | Whether to translate syslog structured data to loki labels. | `false` | no
+`labels`                 | `map(string)` | The labels to associate with each received syslog record. | `{}` | no
+`use_incoming_timestamp` | `bool`        | Whether to set the timestamp to the incoming syslog record timestamp. | `false` | no
+`use_rfc5424_message`    | `bool`        | Whether the full RFC5424-formatted syslog message should be forwarded. | `false` | no
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.syslog.md:74]:
```suggestion
`use_rfc5424_message`    | `bool`        | Whether to forward the full RFC5424-formatted syslog message. | `false` | no
```

File: docs/sources/flow/reference/components/loki.source.syslog.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+`max_message_length`     | `int`         | The maximum limit to the length of syslog messages. | `8192` | no
+
+By default, the component will assign the log entry timestamp as the time it
+was processed.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.syslog.md:78]:
```suggestion
By default, the component assigns the log entry timestamp as the time it
was processed.
```

File: docs/sources/flow/reference/components/loki.source.syslog.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+The `labels` map will be applied to every message that the component reads.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.syslog.md:80]:
```suggestion
The `labels` map is applied to every message that the component reads.
```

File: docs/sources/flow/reference/components/loki.source.syslog.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+All header fields from the parsed RFC5424 messages will be brought in as
+internal labels, prefixed with `__syslog_`.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.syslog.md:83]:
```suggestion
All header fields from the parsed RFC5424 messages are brought in as
internal labels, prefixed with `__syslog_`.
```

File: docs/sources/flow/reference/components/loki.source.syslog.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+If `label_structured_data` is set, structured data in the syslog header will
+also be translated to internal labels in the form of
+`__syslog_message_sd_<ID>_<KEY>`. A structured data entry of `[example@99999
+test="yes"]` would become the label `__syslog_message_sd_example_99999_test`
+with the value `"yes"`.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.syslog.md:89]:
```suggestion
If `label_structured_data` is set, structured data in the syslog header are
also translated to internal labels in the form of
`__syslog_message_sd_<ID>_<KEY>`. For example, a structured data entry of `[example@99999
test="yes"]` becomes the label `__syslog_message_sd_example_99999_test`
with the value `"yes"`.
```

File: docs/sources/flow/reference/components/loki.source.syslog.md (hunk 2 above)
Comment (tpaschalis) [reply docs/sources/flow/reference/components/loki.source.syslog.md:68]: I initially had it this way, but found it was a little stuttering when configuring the top-level block.
```
	listener {
		listen_address  = "127.0.0.1:51898"
		listen_protocol = "udp"
```

File: docs/sources/flow/reference/components/loki.source.syslog.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+## Debug metrics
+* `loki_source_syslog_entries_total` (counter): Total number of successful entries sent to the syslog component.
+* `loki_source_syslog_parsing_errors_total` (counter): Total number of parsing errors while receiving syslog messages.
+* `loki_source_syslog_empty_messages_total` (counter): Total number of empty messages received from the syslog component.
+
+## Example
+
+This example collects listens for Syslog messages in valid RFC5424 format over
+TCP and UDP in the specified ports and forwards them to a `loki.write`
+component.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.syslog.md:120]:
```suggestion
This example listens for Syslog messages in valid RFC5424 format over
TCP and UDP in the specified ports and forwards them to a `loki.write`
component.
```

**Review by karengermond — APPROVED** [review]
LGTM, a few comments

File: docs/sources/flow/reference/components/loki.source.syslog.md (hunk 1 above)
Comment (tpaschalis) [reply docs/sources/flow/reference/components/loki.source.syslog.md:60]: Thanks! And 🤦 about the future tense _once more_, I still haven't gotten used to it.

Comment (karengermond) [reply docs/sources/flow/reference/components/loki.source.syslog.md:60]: > Thanks! And 🤦 about the future tense _once more_, I still haven't gotten used to it.

Haha, no worries!

**Comment by tpaschalis** [comment]
Apologies, forgot to add tests for the component itself, putting it back in draft for a bit.

File: docs/sources/flow/reference/components/loki.source.syslog.md (hunk 2 above)
Comment (mattdurham) [reply docs/sources/flow/reference/components/loki.source.syslog.md]: Fine with that. That aligns with our command line flags which use address.

**Review by rfratto — APPROVED** [review]
Looks good, I don't think I need to review the tests once you get those in 👍

File: docs/sources/flow/reference/components/loki.source.syslog.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+expects syslog messages to be sent to, as well as its behavior when receiving
+messages.
+
+The following arguments can be used to configure a `listener`. Only the
+`address` field is required and any omitted fields take their default
+values.
+
+Name                     | Type          | Description | Default | Required
+------------------------ | ------------- | ----------- | ------- | --------
+`address`                | `string`      | The `<host:port>` address to listen to for syslog messages. | | yes
```
Comment (rfratto) [inline docs/sources/flow/reference/components/loki.source.syslog.md:68]: Should be using placeholder syntax here (so people don't think `<>` is something they need to type):

```suggestion
`address`                | `string`      | The `HOST:PORT` address to listen to for syslog messages. | | yes
```

Comment (tpaschalis) [reply docs/sources/flow/reference/components/loki.source.syslog.md:68]: Thanks for the catch!

**Review by mattdurham — APPROVED** [review]
LGTM

**Comment by rfratto** [comment]
> To test this component, I've used a configuration and a command like the ones presented below. I'm still not able to send messages over the UDP connection as I'm getting msg="error parsing syslog stream" err="invalid or unsupported framing. first byte: 'X'" even for valid RFC5424 entries.

Do you get the same thing when using Promtail? If you get the same issue with that I'd say it's probably fine to leave as-is for now. Otherwise, we might want to look into it more.

**Comment by tpaschalis** [comment]
> Do you get the same thing when using Promtail? If you get the same issue with that I'd say it's probably fine to leave as-is for now. Otherwise, we might want to look into it more.

Last time out yes, I couldn't make UDP work over the command-line. In tests, the UDP messages are received and parsed correctly, I'll check again if I'm doing anything wrong when constructing the message.

**Comment by tpaschalis** [comment]
Hey team! After merging #2697, I've pushed a new commit  https://github.com/grafana/agent/pull/2674/commits/62ee226c4acfd869a74ff4938a811ddab5f56da1 which enables users to wire in relabel rules and make use of the internal labels.

Wanna have another look? cc @mattdurham @rfratto

**Review by rfratto — APPROVED** [review]
👌 I like it. I'm cool with this being merged provided you tested the relabel capsule and everything seems to work ok.

**Pull request merged**